### PR TITLE
Simulation support time precision

### PR DESCRIPTION
**Pull request by kleinai**
- [X] Query simulators for time precision
- [x] Add sim options for time scale and precision override (if applicable)
- [x] Add time precision options to sim functions which need them

### CODE REVIEW

File: core/src/main/scala/spinal/core/sim/SimBootstraps.scala [hunk 1]
```diff
@@ -237,7 +242,8 @@ case class SpinalGhdlBackendConfig[T <: Component](override val rtl : SpinalRepo
                                                    override val simulatorFlags    : ArrayBuffer[String] = ArrayBuffer[String](),
                                                    override val usePluginsCache   : Boolean = true,
                                                    override val pluginsCachePath  : String = "./simWorkspace/.pluginsCachePath",
-                                                   override val enableLogging     : Boolean = false) extends
+                                                   override val enableLogging     : Boolean = false,
+                                                   override val timeScale         : TimeNumber = TimeNumber(BigDecimal("1e-9"))) extends
```
Comment (Dolu1990) [inline core/src/main/scala/spinal/core/sim/SimBootstraps.scala:246]: Normaly, 1 ns should be equivalent to TimeNumber(BigDecimal("1e-9")) :)

Comment (Dolu1990) [reply core/src/main/scala/spinal/core/sim/SimBootstraps.scala:246]: i mean "1 ns"   as code

**Comment by Dolu1990** [comment]
Thanks ^^
Should't we go for 1 ps as default timescale ? i mean it is realy easy to be bounded by 1 ns resolution.

File: core/src/main/scala/spinal/core/sim/package.scala [hunk 2]
```diff
@@ -790,6 +795,12 @@ package object sim {
       fork(doStimulus(period))
     }
 
+    def forkStimulus(period: TimeNumber): Unit = {
+      val cyclesScaled = period.toBigDecimal / SimManagerContext.current.manager.timePrecision
+      println("Cycles: " + cyclesScaled)
```
Comment (Dolu1990) [inline core/src/main/scala/spinal/core/sim/package.scala]: lost prinln :)

File: sim/src/main/scala/spinal/sim/VerilatorBackend.scala [hunk 3]
```diff
@@ -341,6 +343,10 @@ JNIEXPORT jboolean API JNICALL ${jniPrefix}eval_1${uniqueId}
    return Verilated::gotFinish();
 }
 
+JNIEXPORT jboolean API JNICALL ${jniPrefix}get_time_precision_1${uniqueId}
```
Comment (Dolu1990) [inline sim/src/main/scala/spinal/sim/VerilatorBackend.scala:346]: not jboolean but jint :D

**Comment by Dolu1990** [comment]
Hi @kleinai 

I just tried : 

```scala
object SimTimeScale {
  def main(args: Array[String]): Unit = {
    SimConfig.doSim(new Component{
      val a = in UInt(16 bits)
      val x = RegNext(a)
    }){ dut =>
      dut.getAllIo.filter(_.isInput).foreach(_.randomize())
      dut.clockDomain.forkStimulus(10)
      for(i <- 0 until 20){
        dut.a #= i
        dut.clockDomain.waitSampling()
      }
    }
  }
}
```

But it fire a : 
```scala
Exception in thread "main" java.lang.UnsatisfiedLinkError: 'int wrapper_verilator.VerilatorNative.get_time_precision_1(long)'
	at wrapper_verilator.VerilatorNative.get_time_precision_1(Native Method)
	at wrapper_verilator.VerilatorNative.get_time_precision(VerilatorNative.java:7)
	at spinal.sim.SimVerilator.getTimePrecision(SimVerilator.scala:106)
	at spinal.sim.SimManager.<init>(SimManager.scala:108)
	at spinal.core.sim.SimCompiled$$anon$2.<init>(SimBootstraps.scala:539)
	at spinal.core.sim.SimCompiled.doSimApi(SimBootstraps.scala:539)
	at spinal.core.sim.SimCompiled.doSim(SimBootstraps.scala:514)
	at spinal.core.sim.SpinalSimConfig.doSim(SimBootstraps.scala:838)
	at spinal.tester.code.SimTimeScale$.main(Debug.scala:1331)
	at spinal.tester.code.SimTimeScale.main(Debug.scala)

```

I tried stuff to patch it up, but i realy don't know why jni isn't happy XD 
(i just found a jboolean instead of jint somewere, but doesn't fix it XD)

Do you have the same issue ?

Thanks ^^

**Comment by kleinai** [comment]
Compile bug fixed. It took an embarrassingly long time to figure out the extra underscores were the problem.

**Comment by Dolu1990** [comment]
> Compile bug fixed. It took an embarrassingly long time to figure out the extra underscores were the problem.

Ahhhhh my god XD
Yes, i think i spent like 40 mn trying to find the issue XDXD

Thanks :D

**Comment by Dolu1990** [comment]
What's about the set a default timescale / prescision to 1 ps / 1 ps ^^ ?

**Comment by kleinai** [comment]
Should sim sleeps in cycles use the time scale? Shouldn't we default to 1 ns to match older expected behavior?

I'd prefer to keep time precision undefined and unforced. There's a hang up where the mcode GHDL simulator won't run at any precision except 1 fs.

**Comment by Dolu1990** [comment]
@kleinai I just added a centralized timeToLong function to convert stuff.

I think it is all good, but i can't test anything for the closed source pay-wall simulator. did you had a chance to test some ?

Thanks for the PR :D

**Comment by Dolu1990** [comment]
I think you are right we can keep things with their default. I just pushed a fix for the verilator backend which was doing a null exception

**Comment by kleinai** [comment]
I also only have the open source simulators. I roughly put in time precision support for VCS. Looks like the flag was already in there.

**Comment by Dolu1990** [comment]
hmm I'm a big worry that it may break stuff if not tested with VCS / XSIM.

@name1e5s Hi ^^, can you give a try ? Those two backend may be affected.


```scala
object SimTimeScale {
  def main(args: Array[String]): Unit = {
    SimConfig.with???.withWave.doSim(new Component{ //<-   withVcs   withXSim
      val a = in UInt(16 bits)
      val x = RegNext(a)
    }){ dut =>
      dut.getAllIo.filter(_.isInput).foreach(_.randomize())
      dut.clockDomain.forkStimulus(10 us)  // <- in the generated wave file, clock should run at 10 us 
      for(i <- 0 until 20){
        dut.a #= i
        dut.clockDomain.waitSampling()
      }
    }
  }
}
```

**Comment by kleinai** [comment]
I have tested this on XSim because it's free and I'm a Xilinx fan. :P
That's actually where I first started work on it.

**Comment by Dolu1990** [comment]
@kleinai Hooo nice :D

Let's gooooo

Thanks !

**Comment by Dolu1990** [comment]
So can i merge ? (the PR is still "Draft")

**Comment by Dolu1990** [comment]
Thanks :D

**Comment by Dolu1990** [comment]
@kleinai 

Seems the timescale support break for iverilog and vcs for some people XD

VCS => 
An exception or error caused a run to abort: spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(JLspinal/sim/vpi/SharedMemIface;)I 
java.lang.UnsatisfiedLinkError: spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(JLspinal/sim/vpi/SharedMemIface;)I
    at spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(Native Method)
    at spinal.sim.vpi.SharedMemIface.get_time_precision(SharedMemIface.java:108)
    at spinal.sim.SimVpi.getTimePrecision(SimVpi.scala:133)

IVerilog => 

[Progress] IVerilog compilation started
[Progress] IVerilog compilation done in 71.589 ms
Exception in thread "main" java.lang.UnsatisfiedLinkError: 'int spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(long, spinal.sim.vpi.SharedMemIface)'
    at spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(Native Method)
    at spinal.sim.vpi.SharedMemIface.get_time_precision(SharedMemIface.java:108)
    at spinal.sim.SimVpi.getTimePrecision(SimVpi.scala:133)

**Comment by hanm2019** [comment]
> @kleinai
> 
> Seems the timescale support break for iverilog and vcs for some people XD
> 
> VCS => An exception or error caused a run to abort: spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(JLspinal/sim/vpi/SharedMemIface;)I java.lang.UnsatisfiedLinkError: spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(JLspinal/sim/vpi/SharedMemIface;)I at spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(Native Method) at spinal.sim.vpi.SharedMemIface.get_time_precision(SharedMemIface.java:108) at spinal.sim.SimVpi.getTimePrecision(SimVpi.scala:133)
> 
> IVerilog =>
> 
> [Progress] IVerilog compilation started [Progress] IVerilog compilation done in 71.589 ms Exception in thread "main" java.lang.UnsatisfiedLinkError: 'int spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(long, spinal.sim.vpi.SharedMemIface)' at spinal.sim.vpi.JNISharedMemIfaceJNI.SharedMemIface_get_time_precision(Native Method) at spinal.sim.vpi.SharedMemIface.get_time_precision(SharedMemIface.java:108) at spinal.sim.SimVpi.getTimePrecision(SimVpi.scala:133)

I get the same error on VCS 2018, the spinalHDL version is 1.8.0, Does this bug fixed?

**Comment by kleinai** [comment]
See this comment here. I believe you may have the same issue.
https://github.com/SpinalHDL/SpinalHDL/issues/973#issuecomment-1407679128

**Comment by hanm2019** [comment]
Thanks, I solve it by delete the `simWorkspace/.pluginsCachePath`

**Pull request merged**